### PR TITLE
feat: Merge onion_announce changes from new groupchats fork

### PR DESCRIPTION
**Pull request by JFreegman**
These changes will make the DHT compatible with the new groupchats implementation. Additionally, onion announces now support arbitrary extra data.
 

This change is [Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2164)

### CODE REVIEW

**Review by sudden6 — CHANGES_REQUESTED** [review]
> Reviewed 11 of 11 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2164)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @JFreegman)

___
*[toxcore/onion_announce.h, line 26 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybVlF-CJSwAY90CHGG:-MybVlF-CJSwAY90CHGH:b-ms0zq6) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.h#L26)):*
> ```objc
> 
> #define ONION_ANNOUNCE_RESPONSE_MIN_SIZE (2 + ONION_ANNOUNCE_SENDBACK_DATA_LENGTH + CRYPTO_NONCE_SIZE + ONION_PING_ID_SIZE + CRYPTO_MAC_SIZE)
> #define ONION_ANNOUNCE_RESPONSE_MAX_SIZE (ONION_ANNOUNCE_RESPONSE_MIN_SIZE +  ONION_MAX_EXTRA_DATA_SIZE * MAX_SENT_NODES)
> ```

extra space after `+`
___
*[toxcore/onion_announce.h, line 126 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybW8-J0xAB_359HFAH:-MybW8-J0xAB_359HFAI:b-5iv4hf) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.h#L126)):*
> ```objc
>                                uint8_t *response, uint16_t response_size, uint16_t offset);
> 
> // TODO(Jfreegman): Uncomment this when we merge the rest of new groupchats
> ```

General question, why even merge this commented out code, if it's only needed later? Here and other instances
___
*[toxcore/onion_announce.c, line 405 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybW_pc6PMTx5T7OntJ:-MybW_pc6PMTx5T7OntK:b-1lkqz9) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L405)):*
> ```C
> {
>     if (index < 0) {
>         response[0] = 0;
> ```

can you turn `0` into a named constant? Ideally the value would also be a named constant.
___
*[toxcore/onion_announce.c, line 438 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybX0IR8F2cY2rK0Vix:-MybX0IR8F2cY2rK0Viy:b-owvegm) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L438)):*
> ```C
>  *   extra data callback.
>  * @param pack_extra_data_callback Callback that may write extra data into the packet.
>  */
> ```

Please document the return values, alternatively I think the return type could be changed to `bool`.
___
*[toxcore/onion_announce.c, line 506 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybXbtk3I7U-q51xrUa:-MybXbtk3I7U-q51xrUb:buh8itj) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L506)):*
> ```C
>     if (num_nodes != 0) {
>         nodes_length = pack_nodes(onion_a->log, response + nodes_offset, sizeof(nodes_list), nodes_list,
>                                   (uint16_t)num_nodes);
> ```

ensure the cast to uint16_t is safe.



_Code quote:_
```C
num_nodes
```
___
*[toxcore/onion_announce.c, line 519 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybXkDZ5W_53GgJ-f1t:-MybXkDZ5W_53GgJ-f1u:bbwsmhx) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L519)):*
> ```C
> 
>     if (want_node_count) {
>         response[1 + ONION_PING_ID_SIZE] = (uint8_t)num_nodes;
> ```

here it's even casted to an `uint8_t`
___
*[toxcore/onion_announce.c, line 522 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYJSJ1dbItkYZjl9D:-MybYJSJ1dbItkYZjl9E:b-ivp4c5) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L522)):*
> ```C
>     }
> 
>     const int extra_size = pack_extra_data_callback == nullptr ? 0 : pack_extra_data_callback(
> ```

nit: this code is quite hard to read, if it's possible to split that better without losing the const I'd prefer that.
___
*[toxcore/onion_announce.c, line 565 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYjlk3O-hzNKUInLD:-MybYjlk3O-hzNKUInLE:b741xg5) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L565)):*
> ```C
> 
> non_null()
> static int handle_gca_announce_request(Onion_Announce *onion_a, const IP_Port *source, const uint8_t *packet,
> ```

possibly could be changed to return `bool`
___
*[toxcore/onion_announce.c, line 582 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYuMs-XwYjNtIDt_h:-MybYuMs-XwYjNtIDt_i:bb7od0r) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L582)):*
> ```C
> 
> non_null(1, 2, 3) nullable(5)
> static int handle_announce_request(void *object, const IP_Port *source, const uint8_t *packet, uint16_t length,
> ```

same
___
*[toxcore/onion_announce.c, line 598 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYxAZ1k0DmBomUp7a:-MybYxAZ1k0DmBomUp7b:bb7od0r) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L598)):*
> ```C
> /* TODO(Jfreegman): DEPRECATE */
> non_null(1, 2, 3) nullable(5)
> static int handle_announce_request_old(void *object, const IP_Port *source, const uint8_t *packet, uint16_t length,
> ```

same

**Review by JFreegman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2164)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @sudden6)

___
*[toxcore/onion_announce.h, line 26 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybVlF-CJSwAY90CHGG:-Mybmns62_rVsuB8F89c:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.h#L26)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

extra space after `+`
</blockquote></details>

Done.
___
*[toxcore/onion_announce.h, line 126 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybW8-J0xAB_359HFAH:-MybmoTQ1tcNBSLB1vJN:blbvdmh) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.h#L126)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

General question, why even merge this commented out code, if it's only needed later? Here and other instances
</blockquote></details>

It just makes things easier for future me
___
*[toxcore/onion_announce.c, line 405 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybW_pc6PMTx5T7OntJ:-MybwnV-9mTQjdYXoMzF:bn6edmn) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L405)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

can you turn `0` into a named constant? Ideally the value would also be a named constant.
</blockquote></details>

I agree but I think it should be done in a separate commit after the merge. A lot of the onion code can/should be cleaned up and refactored but I think it's a bit outside the scope of the NGC merge.
___
*[toxcore/onion_announce.c, line 438 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybX0IR8F2cY2rK0Vix:-MybxVSD-s002nxQsL2J:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L438)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Please document the return values, alternatively I think the return type could be changed to `bool`.
</blockquote></details>

Done.
___
*[toxcore/onion_announce.c, line 506 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybXbtk3I7U-q51xrUa:-MybxvIYCaXV7LeZ2go3:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L506)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

ensure the cast to uint16_t is safe.
</blockquote></details>

Done.
___
*[toxcore/onion_announce.c, line 519 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybXkDZ5W_53GgJ-f1t:-MybxwPQDFOcZ-HPmotZ:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L519)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

here it's even casted to an `uint8_t`
</blockquote></details>

Done.
___
*[toxcore/onion_announce.c, line 522 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYJSJ1dbItkYZjl9D:-MybyH9X2-25zQU6L_1a:bksyumw) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L522)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

nit: this code is quite hard to read, if it's possible to split that better without losing the const I'd prefer that.
</blockquote></details>

Done. I tried my best but astyle makes it ugly again.
___
*[toxcore/onion_announce.c, line 565 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYjlk3O-hzNKUInLD:-MybyRArEbKc_JRqlV8b:blxt5oy) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L565)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

possibly could be changed to return `bool`
</blockquote></details>

All three of these callbacks could/should be changed to bool, but it would require a larger refactor which goes back to my previous comment. I think that belongs in a separate cleanup commit since it's mostly unrelated to NGC.
___
*[toxcore/onion_announce.c, line 582 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYuMs-XwYjNtIDt_h:-Mybydxz-zBzdcDZ4yjI:b-q64pkt) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L582)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

same
</blockquote></details>

See above.
___
*[toxcore/onion_announce.c, line 598 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYxAZ1k0DmBomUp7a:-MybyeiuCIlT7BEqOzpN:b-q64pkt) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L598)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

same
</blockquote></details>

See above.

File: toxcore/onion_announce.c
Comment (iphydf) [inline toxcore/onion_announce.c:62]: You could keep it in, make it `non_null(1) nullable(2, 3)` and call it from the `new` function with `nullptr`s.

**Review by iphydf — CHANGES_REQUESTED** [review]

File: toxcore/onion_announce.c
Comment (JFreegman) [reply toxcore/onion_announce.c]: Done

**Review by sudden6 — CHANGES_REQUESTED** [review]
> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2164)* status: 2 change requests, 0 of 1 approvals obtained (waiting on @iphydf and @JFreegman)

___
*[toxcore/onion_announce.c, line 405 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybW_pc6PMTx5T7OntJ:-MycBiQkB_R6cqkSheQw:b-60car2) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L405)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary><blockquote>

I agree but I think it should be done in a separate commit after the merge. A lot of the onion code can/should be cleaned up and refactored but I think it's a bit outside the scope of the NGC merge.
</blockquote></details>

Ok then.
___
*[toxcore/onion_announce.c, line 565 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYjlk3O-hzNKUInLD:-MycCHvgCpxkDZhH5fy5:bb9frx8) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L565)):*
<details><summary><i>Previously, JFreegman wrote…</i></summary><blockquote>

All three of these callbacks could/should be changed to bool, but it would require a larger refactor which goes back to my previous comment. I think that belongs in a separate cleanup commit since it's mostly unrelated to NGC.
</blockquote></details>

Could you make a PR or at least open an issue?

**Review by JFreegman — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2164)* status: 2 change requests, 0 of 1 approvals obtained (waiting on @iphydf and @sudden6)

___
*[toxcore/onion_announce.c, line 565 at r1](https://reviewable.io/reviews/toktok/c-toxcore/2164#-MybYjlk3O-hzNKUInLD:-MycVIBf8o5lFOGrejHr:b-896fix) ([raw file](https://github.com/toktok/c-toxcore/blob/7a21b8a9e39148c4b78b66a1d688fc8bc11a626b/toxcore/onion_announce.c#L565)):*
<details><summary><i>Previously, sudden6 wrote…</i></summary><blockquote>

Could you make a PR or at least open an issue?
</blockquote></details>

Done.

**Review by sudden6 — APPROVED** [review]
> Reviewed 2 of 2 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/toktok/c-toxcore/2164)* status: 1 change requests, 0 of 1 approvals obtained (waiting on @iphydf)

**Review by iphydf — APPROVED** [review]

**Comment by zugz** [comment]
I read this and had a few comments; please see HEAD of 
zugz/c-toxcore:ngc_onion_announce (I can try to put this into Reviewable 
or whatever if you prefer).

**Pull request merged**
